airbyte-integrations/connectors/source-marketo/source_marketo/source.py

```diff
@@ -0,0 +1,530 @@
+        params.update(
+            {
+                "maxReturn": self.page_size,
+                "earliestUpdatedAt": stream_slice["startAt"],
+                "latestUpdatedAt": stream_slice["endAt"],
+            }
+        )
+
+        return params
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        for record in super().parse_response(response, stream_state, **kwargs):
+            record["updatedAt"] = record["updatedAt"][:-5]
+            yield record
+
+
+class Campaigns(IncrementalMarketoStream):
+    """
+    Return list of all campaigns.
+    API Docs: http://developers.marketo.com/rest-api/endpoint-reference/lead-database-endpoint-reference/#!/Campaigns/getCampaignsUsingGET
+    """
+
+    pass
+
+
+class Lists(IncrementalMarketoStream):
+    """
+    Return list of all lists.
+    API Docs: http://developers.marketo.com/rest-api/endpoint-reference/lead-database-endpoint-reference/#!/Static_Lists/getListsUsingGET
+    """
+
+    pass
+
+
+class MarketoAuthenticator(Oauth2Authenticator):
+    def __init__(self, config):
+        super().__init__(
+            token_refresh_endpoint=f"{config['base_url']}/identity/oauth/token",
+            client_id=config["client_id"],
+            client_secret=config["client_secret"],
+            refresh_token=None,
+        )
+
+    def get_refresh_request_params(self) -> Mapping[str, Any]:
+        payload: MutableMapping[str, Any] = {
+            "grant_type": "client_credentials",
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+
+        return payload
+
+    def refresh_access_token(self) -> Tuple[str, int]:
+        """
+        Returns a tuple of (access_token, token_lifespan_in_seconds)
+        """
+        try:
+            response = requests.request(method="GET", url=self.token_refresh_endpoint, params=self.get_refresh_request_params())
+            response.raise_for_status()
+            response_json = response.json()
+            # print(response_json["access_token"])
+            return response_json["access_token"], response_json["expires_in"]
+        except Exception as e:
+            raise Exception(f"Error while refreshing access token: {e}") from e
+
+
+class SourceMarketo(AbstractSource):
+    """
+    Source Marketo fetch data of personalized multi-channel programs and campaigns to prospects and customers.
+    """
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        """
+        Testing connection availability for the connector by granting the credentials.
+        """
+
+        try:
+            url = f"{config['base_url']}/rest/v1/leads/describe"
+
+            authenticator = MarketoAuthenticator(config)
+
+            session = requests.get(url, headers=authenticator.get_auth_header())
+            session.raise_for_status()
+
```

Comment:
Is there some opportunity for checking of possible permissions?

Reply:
No.